### PR TITLE
Hangman: Support saving hangman entries

### PR DESCRIPTION
**Pull request by mia-pi-git**
Requested.

### CODE REVIEW

File: server/chat-plugins/hangman.ts [hunk 1]
```diff
@@ -3,6 +3,15 @@
 * By bumbadadabum and Zarel. Art by crobat.
 */
 import {Utils} from '../../lib/utils';
+import {FS} from '../../lib/fs';
+
+const HANGMAN_FILE = 'config/chat-plugins/hangman.json';
+
+export let hangmanData: {[roomid: string]: {[question: string]: string[]}} = {};
```
Comment (KrisXV) [inline server/chat-plugins/hangman.ts:10]:
```suggestion
export let hangmanData: {[roomid: string]: {[phrase: string]: string[]}} = {};
```

this should be formatted as
```javascripts
roomid: {
	"this is a phrase": ["this is a possible hint", "this is another hint for the same phrase"],
	"this is another phrase": ["this is a possible hint", "this is another hint for the same phrase"],
},
```

File: server/chat-plugins/hangman.ts [hunk 2]
```diff
@@ -299,6 +327,86 @@ export const commands: ChatCommands = {
...
+			this.modlog(`HANGMAN RANDOM`);
+		},
+		addrandom(target, room, user) {
+			room = this.requireRoom();
+			this.checkCan('mute', null, room);
+			if (!hangmanData[room.roomid]) hangmanData[room.roomid] = {};
+			if (!target) return this.parse('/help hangman');
+			// validation
+			const args = target.split(',');
+			const {word} = Hangman.validateParams(args);
```
Comment (KrisXV) [inline server/chat-plugins/hangman.ts:351]:
```suggestion
			const {phrase} = Hangman.validateParams(args);
```

File: server/chat-plugins/hangman.ts [hunk 3]
```diff
@@ -299,6 +327,86 @@ export const commands: ChatCommands = {
...
+			game.display(user, true);
+			this.modlog(`HANGMAN RANDOM`);
+		},
+		addrandom(target, room, user) {
+			room = this.requireRoom();
+			this.checkCan('mute', null, room);
+			if (!hangmanData[room.roomid]) hangmanData[room.roomid] = {};
+			if (!target) return this.parse('/help hangman');
+			// validation
+			const args = target.split(',');
```
Comment (KrisXV) [inline server/chat-plugins/hangman.ts:350]:
```suggestion
			const args = target.split('|');
```

some people might want hints with commas in them

**Review by KrisXV — CHANGES_REQUESTED** [review]
Unless I misinterpreted the suggestion, I think the requester wanted to be able to synchronously create a hangman and save it to a database for future use

**Comment by mia-pi-git** [comment]
> Unless I misinterpreted the suggestion, I think the requester wanted to be able to synchronously create a hangman and save it to a database for future use

They said they wanted it the same as kid a - which saved them without creating one. https://www.smogon.com/forums/threads/hangmans-save.3671044/

**Review by KrisXV — APPROVED** [review]

**Comment by mia-pi-git** [comment]
@Zarel bump

**Comment by Zarel** [comment]
:s

I suppose the memory impact isn't big...

File: server/chat-plugins/hangman.ts [hunk 4]
```diff
@@ -205,6 +214,37 @@ export class Hangman extends Rooms.RoomGame {
...
+	static save() {
+		FS(HANGMAN_FILE).writeUpdate(() => JSON.stringify(hangmanData));
+	}
+	static getRandom(room: RoomID) {
+		if (!hangmanData[room]) {
+			hangmanData[room] = {};
+			this.save();
+			throw new Chat.ErrorMessage(`The room ${room} has no saved hangman words.`);
+		}
+		const shuffled = Utils.shuffle(Object.keys(hangmanData[room]))[0];
```
[the remark this answers is not in the document]
Comment (Zarel) [reply server/chat-plugins/hangman.ts:226]:
```suggestion
		const shuffled = Utils.randomElement(Object.keys(hangmanData[room]));
```

File: server/chat-plugins/hangman.ts [hunk 5]
```diff
@@ -205,6 +214,37 @@ export class Hangman extends Rooms.RoomGame {
...
+		const hints = hangmanData[room][shuffled];
+		return {
+			question: shuffled,
+			hint: Utils.shuffle(hints)[0],
```
Comment (Zarel) [inline server/chat-plugins/hangman.ts:230]:
```suggestion
			hint: Utils.randomElement(hints),
```

File: lib/utils.ts [hunk 6]
```diff
@@ -220,6 +220,10 @@ export const Utils = new class Utils {
 		return arr;
 	}
 
+	randomElement<T>(arr: T[]): T {
+		return this.shuffle(arr)[0];
+	}
```
Comment (Zarel) [inline lib/utils.ts:225]: Please no!

Comment (Zarel) [reply lib/utils.ts:225]:
```suggestion
	randomElement<T>(arr: T[]): T {
		const i = Math.floor(Math.random() * arr.length);
		return arr[i];
	}
```

Comment (mia-pi-git) [reply lib/utils.ts:225]: lol this is what i did originally

**Comment by Zarel** [comment]
ty!

**Pull request merged**